cirq-core/cirq/devices/noise_utils.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Sequence
+import warnings
+import numpy as np
+
+if TYPE_CHECKING:
+    import cirq
+
+
+# Tag for gates to which noise must be applied.
+PHYSICAL_GATE_TAG = 'physical_gate'
+
+
+@dataclass(frozen=True)
+class OpIdentifier:
+    """Identifies an operation by gate and (optionally) target qubits."""
+
+    gate_type: type
```

Comment:
I would be wary of making gate_type a 'type' and not a primitive type like str. If you ever want to be able to serialize an OpIdentifier, a primitive type could make your life much easier. What if a user wanted an OpIdentifier for a gate that was not serializable?

Reply:
(As noted elsewhere, `OpIdentifier` will be superseded by the use of `cirq.Gateset`. The comment below is just to clarify the previous behavior.)

The current way I handle this is by converting gate types to "canonical" instantiations of those types, which can be serialized (see #4640). Since Cirq requires that Circuits are serializable, all gates are expected to also be serializable.

I preferred types over strings for this implementation because they allow us to rely on the gate type hierarchy for matching behavior. A `SycamoreGate` should match both `FSimGate` and `SycamoreGate` mappings, but if both are present it should prefer `SycamoreGate` since it's more precise.

Reply:
`OpIdentifier` has been modified to use `cirq.GateFamily` under the hood, which is serializable. (This change occurred in a previous PR in the sequence, and might not be highlighted in this PR.)